fullstory-segment-middleware/src/main/java/com/fullstorydev/fullstory_segment_middleware/FullStorySegmentMiddleware.java

```diff
@@ -159,4 +167,118 @@ private BasePayload getNewPayloadWithFSURL(BasePayload payload, Map<String, Obje
+                        stack.push(m);
+                    }
+                } else if (item != null && item.getClass().isArray()) {
+                    // if it's of array type (but not iterable)
+                    // To comply with FS requirements, flatten the array of objects into a map:
+                    // each item in array becomes a map, with key, and item
+                    // enable search value the array in FS (i.e. searching for one product when array of products are sent)
+                    // then push each item with the same key back to stack
+                    Object[] objs = this.getArrayObjectFromArray(item);
+                    for (Object obj: objs) {
+                        Map<String,Object> m = new HashMap<>();
+                        m.put(key, obj);
+                        stack.push(m);
+                    }
+                } else if (item instanceof Iterable) {
+                    for (Object obj: (Iterable<?>) item) {
+                        Map<String,Object> m = new HashMap<>();
+                        m.put(key, obj);
+                        stack.push(m);
+                    }
+                } else {
+                    // if this is not a map or array, simply treat as a "primitive" value and send them as-is
+                    String suffix = this.getSuffixStringFromSimpleObject(item);
+                    this.appendSimpleObjectToMap(mutableProps, key + suffix, item);
+                }
+            }
+        }
+        pluralizeAllArrayKeysInMap(mutableProps);
+        return mutableProps;
+    }
+
+    private void appendSimpleObjectToMap (Map<String, Object> map, String key, Object obj) {
+        // add one obj into the result map, check if the key with suffix already exists, if so append to the result arrays.
+        // key is already suffixed, and always in singular form
+        Object item = map.get(key);
+        if (item != null) {
+            // if the same key already exist, check if plural key is already in the map
+            // concatenate array and replace
+            ArrayList<Object> arr = new ArrayList<>();
+            arr.add(obj);
+            if (item instanceof Collection) {
+                arr.addAll((Collection<?>) item);
+            } else {
+                arr.add(item);
+            }
+            map.put(key, arr);
+        } else {
+            map.put(key, obj);
+        }
+    }
+
+    private String getSuffixStringFromSimpleObject(Object item) {
+        // default to no suffix;
+        String suffix = "";
+        if ( item instanceof String ) {
+            suffix = "_str";
+        } else if ( item instanceof Number ) {
+            // defaut to real
+            suffix = "_real";
+            if ( item instanceof Integer || item instanceof BigInteger) {
+                suffix = "_int";
+            }
+        } else if (item instanceof Boolean) {
+            suffix = "_bool";
+        } else if (item instanceof Date) {
```

Comment:
do we need to check for the different date types, like java.sql.Date? I am not sure if we auto-handle that anywhere else.

Reply:
Yeah we definitely should, there are some caveats in iOS as well in Dates due to encoding. Good news is that in Segment spec, all dates should already be suffixed with _date, and even if we fail it should still work. 
I think we should parse the date to make sure it fit FS required format too, so this may be a bigger/separate PR. WDTY